OpenRA.Mods.Common/Widgets/Logic/Editor/MapMirrorTilesLogic.cs

```diff
@@ -0,0 +1,341 @@
+			axisAngleValueLabel.IsVisible = IsFlipMode;
+			axisAngleValueLabel.GetText = () => mirrorLayerTrait.AxisAngle.ToString(NumberFormatInfo.InvariantInfo);
+
+			axisAngleGuide = widget.Get<CheckboxWidget>("AXIS_ANGLE_GUIDE");
+			axisAngleGuide.IsVisible = IsFlipMode;
+			axisAngleGuide.IsChecked = () => mirrorLayerTrait.ShowAxisGuide;
+			axisAngleGuide.OnClick = () => mirrorLayerTrait.ShowAxisGuide = !mirrorLayerTrait.ShowAxisGuide;
+		}
+
+		protected override void Dispose(bool disposing)
+		{
+			editor.BrushChanged -= HandleBrushChanged;
+			base.Dispose(disposing);
+		}
+
+		void HandleBrushChanged()
+		{
+			if (editor.CurrentBrush is not EditorMirrorLayerBrush)
+			{
+				mirrorTile = null;
+			}
+		}
+
+		void ClearSelected()
+		{
+			if (editor.CurrentBrush is EditorMirrorLayerBrush mirrorLayerBrush &&
+				mirrorLayerBrush.Template.HasValue &&
+				mirrorLayerTrait.Tiles.TryGetValue(mirrorLayerBrush.Template.Value, out var tiles) &&
+				tiles.Count > 0)
+				editorActionManager.Add(new ClearSelectedMirrorTilesEditorAction(mirrorLayerBrush.Template.Value, mirrorLayerTrait));
+		}
+
+		void ClearAll()
+		{
+			if (mirrorLayerTrait.Tiles.Count > 0 && mirrorLayerTrait.Tiles.Any(x => x.Value.Count > 0))
+				editorActionManager.Add(new ClearAllMirrorTilesEditorAction(mirrorLayerTrait));
+		}
+
+		void ShowMirrorModeDropDown(DropDownButtonWidget dropdown)
+		{
+			ScrollItemWidget SetupItem(MirrorTileMode mode, ScrollItemWidget itemTemplate)
+			{
+				var item = ScrollItemWidget.Setup(itemTemplate,
+					() => mirrorLayerTrait.MirrorMode == mode,
+					() => mirrorLayerTrait.SetMirrorMode(mode));
+
+				item.Get<LabelWidget>("LABEL").GetText = () =>
+				{
+					switch (mode)
+					{
+						case MirrorTileMode.None:
+							return TranslationProvider.GetString(MirrorModeNoneTranslation);
+						case MirrorTileMode.Flip:
+							return TranslationProvider.GetString(MirrorModeFlipTranslation);
+						case MirrorTileMode.Rotate:
+							return TranslationProvider.GetString(MirrorModeRotateTranslation);
+						default:
+							throw new ArgumentException($"Couldn't find translation for mirror tile flip mode '{mode}'");
+					}
+				};
+
+				return item;
+			}
+
+			var options = new[] { MirrorTileMode.None, MirrorTileMode.Flip, MirrorTileMode.Rotate };
+			dropdown.ShowDropDown("LABEL_DROPDOWN_TEMPLATE", 150, options, SetupItem);
+		}
+
+		void ShowFlipNumSidesDropDown(DropDownButtonWidget dropdown)
+		{
+			ScrollItemWidget SetupItem(int value, ScrollItemWidget itemTemplate)
+			{
+				var item = ScrollItemWidget.Setup(itemTemplate,
+					() => mirrorLayerTrait.NumSides == value,
+					() => mirrorLayerTrait.NumSides = value);
+
+				item.Get<LabelWidget>("LABEL").GetText = () => value.ToString(NumberFormatInfo.InvariantInfo);
+				return item;
+			}
+
+			var options = new[] { 2, 4 };
+			dropdown.ShowDropDown("LABEL_DROPDOWN_TEMPLATE", 150, options, SetupItem);
+		}
+
+		class ClearSelectedMirrorTilesEditorAction : IEditorAction
```

Comment:
I don't think that editor actions should be in widget logic. Please move them to the brush